contribs/gmf/src/services/objecteditingquery.js

```diff
@@ -0,0 +1,207 @@
+goog.require('gmf');
+goog.require('gmf.Themes');
+goog.require('ol.format.WMSGetFeatureInfo');
+goog.require('ol.source.ImageWMS');
+
+
+/**
+ * A service that collects all queryable layer nodes from all themes, stores
+ * them and use them to make WMS GetFeatureInfo queries. Queries can be made
+ * regardless of the associated layer visibility. The layer nodes are also
+ * loaded only once.
+ *
+ * @param {angular.$http} $http Angular $http service.
+ * @param {angular.$q} $q Angular $q service.
+ * @param {gmf.Themes} gmfThemes The gmf themes service.
+ * @constructor
+ * @struct
+ * @ngInject
+ */
+gmf.ObjectEditingQuery = function($http, $q, gmfThemes) {
+
+  /**
+   * @type {angular.$http}
+   * @private
+   */
+  this.http_ = $http;
+
+  /**
+   * @type {angular.$q}
+   * @private
+   */
+  this.q_ = $q;
+
+  /**
+   * @type {gmf.Themes}
+   * @private
+   */
+  this.gmfThemes_ = gmfThemes;
+
+  /**
+   * @type {angular.$q.Deferred|null}
+   * @private
+   */
+  this.getQueryableLayerNodesDefered_ = null;
+
+};
+
+
+/**
+ * @return {angular.$q.Promise} Promise.
+ * @export
+ */
+gmf.ObjectEditingQuery.prototype.getQueryableLayersInfo = function() {
+
+  if (!this.getQueryableLayerNodesDefered_) {
+    this.getQueryableLayerNodesDefered_ = this.q_.defer();
+    this.gmfThemes_.getOgcServersObject().then(function(ogcServers) {
+      this.gmfThemes_.getThemesObject().then(function(themes) {
+        if (!themes) {
+          return;
+        }
+
+        var queryableLayersInfo =
+            gmf.ObjectEditingQuery.getQueryableLayersInfoFromThemes(
+              themes,
+              ogcServers
+            );
+
+        // FIXME - missing checking of ObjectEditing config. Do this here...
+        // ...
+
+        this.getQueryableLayerNodesDefered_.resolve(queryableLayersInfo);
+      }.bind(this));
+    }.bind(this));
+  }
+
+  return this.getQueryableLayerNodesDefered_.promise;
+
+};
+
+
+/**
+ * @param {Array.<gmfThemes.GmfTheme>} themes List of theme nodes.
+ * @param {gmfThemes.GmfOgcServers} ogcServers List of ogc servers
+ * @return {Array.<gmf.ObjectEditingQuery.QueryableLayerInfo>} List of
+ *     queryable layers information.
+ */
+gmf.ObjectEditingQuery.getQueryableLayersInfoFromThemes = function(
+  themes, ogcServers
+) {
+  var queryableLayersInfo = [];
+  var theme;
+  var group;
+  var nodes;
+  var node;
+
+  for (var i = 0, ii = themes.length; i < ii; i++) {
+    theme = /** @type {gmfThemes.GmfTheme} */ (themes[i]);
+    for (var j = 0, jj = theme.children.length; j < jj; j++) {
+      group = /** @type {gmfThemes.GmfGroup} */ (theme.children[j]);
+
+      // Skip groups that don't have an ogcServer set
+      if (!group.ogcServer) {
+        continue;
+      }
+
+      nodes = [];
+      gmf.Themes.getFlatNodes(group, nodes);
+
+      for (var k = 0, kk = nodes.length; k < kk; k++) {
+        node = /** @type {gmfThemes.GmfGroup|gmfThemes.GmfLayerWMS} */ (
+          nodes[k]);
+
+        // Skip groups within groups
```

Comment:
I'm not sure, can you really skip them ?
And this kind of method isn't already somewhere ? I'm almost sur that we already collecte layers somewhere...